### PR TITLE
make custom datasource visible from the UI

### PR DESCRIPTION
**Pull request by snopoke**
@czue

### CODE REVIEW

File: corehq/apps/userreports/views.py [hunk 1]
```diff
@@ -12,7 +13,7 @@
 from corehq.apps.domain.decorators import login_and_domain_required
 from corehq.apps.userreports.app_manager import get_case_data_source, get_form_data_source
 from corehq.apps.userreports.exceptions import BadSpecError
-from corehq.apps.userreports.models import ReportConfiguration, DataSourceConfiguration
+from corehq.apps.userreports.models import ReportConfiguration, DataSourceConfiguration, CustomDataSourceConfiguration
```
Comment (dimagimon) [inline corehq/apps/userreports/views.py]: E501 line too long (118 > 115 characters)

File: corehq/apps/userreports/views.py [hunk 2]
```diff
@@ -248,5 +261,5 @@ def _shared_context(domain):
     return {
         'domain': domain,
         'reports': ReportConfiguration.by_domain(domain),
-        'data_sources': DataSourceConfiguration.by_domain(domain),
+        'data_sources': DataSourceConfiguration.by_domain(domain) + list(CustomDataSourceConfiguration.all()),
```
Comment (czue) [inline corehq/apps/userreports/views.py]: would be nice to be able to partition these by domain eventually too.

File: custom/up_nrhm/data_sources/asha_facilitators.json [hunk 3]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (czue) [inline custom/up_nrhm/data_sources/asha_facilitators.json]: did you mean to include this file? it just looks like a copy of the location hierarchy data source.

**Comment by czue** [comment]
looks good pending that one question about the custom file

File: corehq/apps/userreports/views.py (hunk 2 above)
Comment (snopoke) [inline corehq/apps/userreports/views.py]: Ha, good point, shouldn't be to hard to do now.

On 12 February 2015 at 15:30, Cory Zue <notifications@github.com> wrote:

> In corehq/apps/userreports/views.py
> <https://github.com/dimagi/commcare-hq/pull/5520#discussion_r24580004>:
>
> > @@ -248,5 +261,5 @@ def _shared_context(domain):
> >      return {
> >          'domain': domain,
> >          'reports': ReportConfiguration.by_domain(domain),
> > -        'data_sources': DataSourceConfiguration.by_domain(domain),
> > +        'data_sources': DataSourceConfiguration.by_domain(domain) + list(CustomDataSourceConfiguration.all()),
>
> would be nice to be able to partition these by domain eventually too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/5520/files#r24580004>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

File: custom/up_nrhm/data_sources/asha_facilitators.json (hunk 3 above)
Comment (snopoke) [inline custom/up_nrhm/data_sources/asha_facilitators.json]: no

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by snopoke** [comment]
@czue  updated

File: corehq/apps/userreports/models.py [hunk 4]
```diff
@@ -265,15 +265,25 @@ def all(cls):
 
 
 class CustomDataSourceConfiguration(JsonObject):
+    _datasource_id_prefix = 'custom-'
     domains = ListProperty()
     config = DictProperty()
 
     @classmethod
-    def all(cls):
+    def all(cls, domain=None):
```
Comment (czue) [inline corehq/apps/userreports/models.py]: is a bit subjective but I think i would have gone with a `by_domain` classmethod that just calls `.all()` and filters the results. that's more conventional in other parts of HQ, as well as consistent with the `DataSourceConfiguration` API

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

File: corehq/apps/userreports/models.py (hunk 4 above)
Comment (snopoke) [inline corehq/apps/userreports/models.py]: ag, all good points

**Pull request merged**